elastic/kibana
# [Response Ops][Connectors] Add unsecured actions client to allow system to schedule email action

Resolves https://github.com/elastic/kibana/issues/143052

## Summary

Creates an `UnsecuredActionsClient` that contains a single function for bulk enqueuing actions. This differs from the normal bulk enqueue function inside the `ActionsClient` because it does not check RBAC on the user request object in order to verify that the user has access to the `Actions and Connectors` feature. Because we want to allow the system to schedule actions without a user in the middle, we need to bypass the RBAC. This unsecured client uses the `internalSavedObjectsRespository` which uses the internal Kibana user to write the `action_task_params` saved object required to schedule an action.

This MVP limits the available connectors to preconfigured connectors in order to remove spaces from the equation. It also contains two allowlists:
* Allowlist in `x-pack/plugins/actions/server/create_unsecured_execute_function.ts` to limit this feature to just the email connector. Because the system is scheduling this action, we also don't have any API keys to use for the action. The initial notifications feature will only use the email connector which does not require an API key but this allowlist should only contain connectors that don't need API keys (currently only the ES Index connector needs one).
* Allowlist in `x-pack/plugins/actions/server/unsecured_actions_client/unsecured_actions_client.ts` to limit this feature to only certain plugins (`notifications` to start). If another plugin wants to use the unsecured actions client, they should add themselves to the allowlist, which will ping the Response Ops team, allowing us to be aware of where this client is being used.

Note: I added only unit tests in this PR since this functionality will be used by other plugins. I think functional tests should be added by the plugins that use this client.

## To Verify
1. Make sure you have a preconfigured email connector defined in your `kibana.yml`
2. Add `alerting` to the allowlist in `x-pack/plugins/actions/server/unsecured_actions_client/unsecured_actions_client.ts`
3. In `x-pack/plugins/alerting/server/plugin.ts` in the `start` function, add the following:

```
const unsecuredActionsClient = plugins.actions.getUnsecuredActionsClient();
unsecuredActionsClient.bulkEnqueueExecution('alerting', [
  {
    id: <preconfigured_connector_id>,
    params: {
      to: [<email_address>],
      subject: 'hello from Kibana!',
      message: 'does this work??',
    },
  },
]);
```
4. Start Kibana and verify that the email was scheduled and sent.
